[PR-D] 主动追踪 + AI 预验收 + 修订循环

**Part of #4**,依赖 N2 / N3 合并后启动

## 目标
候选人接活后,系统**主动**追踪进度、自动验收代码任务、给出可执行修订反馈。HR 仅在 LLM 拿不准或修订耗尽时被打扰。

## 实现要求

### 1. 主动巡检(APScheduler)
- 每 6 小时扫描所有 `IN_PROGRESS` 任务
- 触发档:50% DDL 时长无进度更新 / DDL 前 24h / DDL 前 2h / DDL 已过
- 50% 档:温和催办卡推候选人
- 24h / 2h 档:升级催办 + HR 收风险卡
- DDL 已过未交付:任务标 `OVERDUE`,HR 收异常卡(可一键改派)

### 2. 进度回报
- 候选人发 `/progress <task_id> <文本>`
- LLM 抽取百分比与风险点写入 task 表
- 主动巡检时若有近 24h 进度则跳过当次催办

### 3. AI 预验收(代码任务专属快车道)
候选人 `/done` 携带 PR URL → 触发链路:

```
PR opened/synchronize ─▶ GitHub Actions 跑 acceptance.yaml ─▶ check_run webhook
                                                                    │
                                                                    ▼
                                                          Taskbot 拉 CI 状态 + diff
                                                                    │
                                                          ┌─────────┴─────────┐
                                                          ▼                   ▼
                                                     全绿 + LLM 高分?    有问题?
                                                          │                   │
                                                          ▼                   ▼
                                                   自动 merge + COMPLETED   修订反馈卡
```

LLM diff review 关注:
- 是否真的回应 issue 描述(防绕过测试)
- N+1 / 安全反模式
- PR 体积 vs `acceptance.yaml` 中的 `max_pr_lines`
- commit message 质量

阈值:env `AUTO_VERIFY_THRESHOLD`(默认 4,5 分制),≥4 自动通过,2-3 给 HR,≤1 自动打回。

### 4. acceptance 契约
N1 建任务时,在 GitHub issue 正文嵌入机器可读的 YAML 块:
```yaml
acceptance:
  ci_required: ["test", "lint", "typecheck"]
  coverage_min: 80
  custom_commands:
    - "pytest tests/integration/test_xxx.py"
  forbid_files: [".env", "*.pem"]
  max_pr_lines: 2000
```
Taskbot 解析此块作为验收依据。

### 5. 修订循环(高 ROI 关卡)
未通过(CI 红 / LLM 找出 issue)时,**结构化反馈卡推候选人**,内容必含 4 类:

| 来源 | 内容 |
|---|---|
| CI 失败摘要 | `pytest XXX FAILED — AssertionError: expected 0, got 3` |
| LLM 评注 | "在 prefilter_by_tags 里调用了 LLM,违反 N1 第 4 节。建议移到 rerank_by_llm" |
| 静态规则违规 | "PR 2147 行 > max_pr_lines=2000,请拆分" |
| 风格提示(不阻塞) | "新增 4 个 print(),建议改用 logger" |

候选人 push 新 commit → 自动重检(`pull_request.synchronize` webhook)。

task 表新增字段:
- `revision_count: int`
- `revision_history: JSON[]`(每轮记 issues + commit_sha + timestamp)

env `MAX_REVISIONS`(默认 3),达上限自动升级给 HR,卡片展示完整修订历史。

### 6. LLM Prompt 强约束
反馈生成 prompt 必须强制每条 issue 包含 `how_to_fix` 字段,缺失则 LLM 重试一次,仍缺失则降级为模板话术。

## 验收(可执行)
```bash
pytest tests/integration/test_proactive_tracking.py -v
# - test_idle_50pct_ddl_triggers_reminder
# - test_ddl_24h_escalates_to_hr
# - test_progress_command_resets_idle_timer
# - test_overdue_marks_status_and_notifies_hr

pytest tests/integration/test_ai_verification.py -v
# - test_green_ci_high_score_auto_merge
# - test_low_score_routes_to_hr
# - test_acceptance_yaml_parsed_from_issue
# - test_max_pr_lines_violation_auto_bounce

pytest tests/integration/test_revision_loop.py -v
# - test_ci_fail_pushes_actionable_card
# - test_feedback_includes_file_line_anchors
# - test_new_commit_triggers_recheck
# - test_revision_count_increments
# - test_max_revisions_escalates_with_history
# - test_how_to_fix_field_always_present
```

--- a/.github/workflows/acceptance.yaml
+++ b/.github/workflows/acceptance.yaml
@@ -1,0 +1,52 @@
+# Acceptance verification workflow — issue #12 §3.
+#
+# This workflow runs on every PR open/update for repos that opt in (the
+# task creation flow embeds an `acceptance:` YAML block in the GitHub
+# issue body — see app/services/acceptance.py). The check_run results
+# land at the Taskbot /webhooks/github endpoint, where the AI reviewer
+# decides auto_merge / hr_review / auto_bounce.
+#
+# Repos that don't use Taskbot can ignore this file.
+name: Acceptance
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened]
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
+          cache: pip
+      - run: pip install -r requirements-dev.txt
+      - run: pytest tests/ --cov=app --cov-report=xml
+      - uses: actions/upload-artifact@v4
+        with:
+          name: coverage-${{ github.sha }}
+          path: coverage.xml
+
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
+          cache: pip
+      - run: pip install -r requirements-dev.txt
+      - run: ruff check .
+
+  typecheck:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
+          cache: pip
+      - run: pip install -r requirements-dev.txt
+      - run: mypy app

--- a/app/config.py
+++ b/app/config.py
@@ -53,6 +53,12 @@ class Settings(BaseSettings):
     reject_fallback_seconds: int = 5
     idempotency_ttl_hours: int = 24
     idempotency_db_path: str = "/tmp/taskbot_idempotency.sqlite3"
+
+    # PR-D: 主动追踪 + AI 预验收 + 修订循环
+    proactive_scan_interval_hours: int = 6
+    auto_verify_threshold: int = 4   # 5 分制,>= 该值自动 merge
+    max_revisions: int = 3           # 修订上限,超过升级 HR
+    progress_idle_grace_hours: int = 24  # 近 24h 有进度则跳过催办
     
     model_config = {
         "env_file": ".env",

--- a/app/services/acceptance.py
+++ b/app/services/acceptance.py
@@ -1,0 +1,220 @@
+"""Acceptance contract — issue #12 §4.
+
+Tasks created via N1 embed a YAML block in their GitHub issue body:
+
+```yaml
+acceptance:
+  ci_required: ["test", "lint", "typecheck"]
+  coverage_min: 80
+  custom_commands:
+    - "pytest tests/integration/test_xxx.py"
+  forbid_files: [".env", "*.pem"]
+  max_pr_lines: 2000
+```
+
+This module parses that block out and returns a typed contract that
+downstream verification (ai_review, revision) consults.
+"""
+
+from __future__ import annotations
+
+import logging
+import re
+from dataclasses import dataclass, field
+from typing import Optional
+
+logger = logging.getLogger(__name__)
+
+# Match a fenced YAML block whose top-level key is `acceptance:` (with or
+# without leading whitespace inside the fence).
+_FENCE_RE = re.compile(
+    r"```ya?ml\s*\n(?P<body>.+?\n)```",
+    re.DOTALL | re.IGNORECASE,
+)
+
+
+@dataclass
+class AcceptanceContract:
+    ci_required: list[str] = field(default_factory=list)
+    coverage_min: int = 0
+    custom_commands: list[str] = field(default_factory=list)
+    forbid_files: list[str] = field(default_factory=list)
+    max_pr_lines: int = 0
+
+    @property
+    def is_default(self) -> bool:
+        """No real contract → conservative defaults (no requirements)."""
+        return (
+            not self.ci_required
+            and self.coverage_min == 0
+            and not self.custom_commands
+            and not self.forbid_files
+            and self.max_pr_lines == 0
+        )
+
+
+def parse_acceptance_yaml(body: str) -> AcceptanceContract:
+    """Extract the ``acceptance:`` YAML block from a GitHub issue body.
+
+    Returns the parsed contract; on missing block / parse error, returns
+    an empty default contract (caller decides how to handle).
+
+    PyYAML is intentionally optional — if not installed we fall back to a
+    minimal hand-rolled parser that handles the documented shape.
+    """
+    if not body:
+        return AcceptanceContract()
+    match = _FENCE_RE.search(body)
+    if not match:
+        logger.debug("acceptance: no YAML fence found")
+        return AcceptanceContract()
+    yaml_text = match.group("body")
+    data = _load_yaml(yaml_text)
+    if not isinstance(data, dict):
+        return AcceptanceContract()
+    payload = data.get("acceptance") or data
+    if not isinstance(payload, dict):
+        return AcceptanceContract()
+
+    return AcceptanceContract(
+        ci_required=_as_list(payload.get("ci_required")),
+        coverage_min=_as_int(payload.get("coverage_min")),
+        custom_commands=_as_list(payload.get("custom_commands")),
+        forbid_files=_as_list(payload.get("forbid_files")),
+        max_pr_lines=_as_int(payload.get("max_pr_lines")),
+    )
+
+
+def _load_yaml(text: str) -> Optional[dict]:
+    """Parse YAML; prefer PyYAML when available."""
+    try:
+        import yaml  # type: ignore
+
+        return yaml.safe_load(text)
+    except ImportError:
+        logger.debug("acceptance: PyYAML missing; using fallback parser")
+        return _fallback_parse(text)
+    except Exception as exc:
+        logger.warning("acceptance: PyYAML parse error: %s", exc)
+        return None
+
+
+def _fallback_parse(text: str) -> dict:
+    """Minimal hand-rolled YAML for the documented schema only.
+
+    Handles:
+    - top-level keys: ``key: value``
+    - nested ``acceptance:`` block via 2-space indent
+    - inline lists ``[a, b]`` and block lists ``- a``
+    - integer auto-coerce
+    """
+    out: dict = {}
+    cur: dict = out
+    section_indent = 0
+    section_key: Optional[str] = None
+    list_buffer: Optional[list] = None
+    list_for_key: Optional[str] = None
+
+    def _flush_list():
+        nonlocal list_buffer, list_for_key
+        if list_buffer is not None and list_for_key is not None:
+            cur[list_for_key] = list_buffer
+        list_buffer = None
+        list_for_key = None
+
+    for raw in text.splitlines():
+        line = raw.rstrip()
+        if not line.strip() or line.lstrip().startswith("#"):
+            continue
+        indent = len(line) - len(line.lstrip())
+
+        if line.lstrip().startswith("- "):
+            if list_buffer is None:
+                continue
+            list_buffer.append(_strip_value(line.lstrip()[2:]))
+            continue
+
+        # Non-list line — flush any pending list
+        _flush_list()
+
+        if indent == 0:
+            # New top-level key
+            cur = out
+            section_indent = 0
+            if ":" in line:
+                key, _, value = line.partition(":")
+                key = key.strip()
+                value = value.strip()
+                if not value:
+                    cur[key] = {}
+                    section_key = key
+                    section_indent = indent
+                    cur = cur[key]
+                else:
+                    out[key] = _coerce(value)
+        elif indent > section_indent:
+            if ":" in line:
+                key, _, value = line.lstrip().partition(":")
+                key = key.strip()
+                value = value.strip()
+                if not value:
+                    list_buffer = []
+                    list_for_key = key
+                else:
+                    cur[key] = _coerce(value)
+        elif indent <= section_indent:
+            cur = out
+            section_indent = 0
+            if ":" in line:
+                key, _, value = line.partition(":")
+                key = key.strip()
+                value = value.strip()
+                if value:
+                    out[key] = _coerce(value)
+                else:
+                    cur[key] = {}
+                    section_key = key
+                    cur = cur[key]
+                    section_indent = indent
+
+    _flush_list()
+    return out
+
+
+_INLINE_LIST_RE = re.compile(r"^\[(.*)\]$")
+
+
+def _coerce(value: str):
+    value = value.strip().strip("\"'")
+    m = _INLINE_LIST_RE.match(value)
+    if m:
+        items = [s.strip().strip("\"'") for s in m.group(1).split(",") if s.strip()]
+        return [_coerce_scalar(i) for i in items]
+    return _coerce_scalar(value)
+
+
+def _coerce_scalar(value: str):
+    if value.isdigit():
+        return int(value)
+    if value.lower() in {"true", "false"}:
+        return value.lower() == "true"
+    return value
+
+
+def _strip_value(value: str):
+    return _coerce_scalar(value.strip().strip("\"'"))
+
+
+def _as_list(value) -> list:
+    if value is None:
+        return []
+    if isinstance(value, list):
+        return [str(v) for v in value]
+    return [str(value)]
+
+
+def _as_int(value) -> int:
+    try:
+        return int(value or 0)
+    except (TypeError, ValueError):
+        return 0

--- a/app/services/ai_review.py
+++ b/app/services/ai_review.py
@@ -1,0 +1,237 @@
+"""AI pre-verification — issue #12 §3.
+
+When a candidate sends ``/done <PR URL>`` the GitHub Actions workflow
+``acceptance.yaml`` runs the ci_required jobs. The check_run webhook
+event lands in :func:`handle_check_run`; we then:
+
+1. Pull CI conclusion + PR diff stats.
+2. Apply the static rules (max_pr_lines, forbid_files, coverage_min).
+3. Ask the LLM to score the PR 1-5 against the issue description.
+4. Route per AUTO_VERIFY_THRESHOLD:
+   - score >= threshold AND CI green → auto merge + COMPLETED
+   - 2 <= score < threshold      → send to HR for manual decision
+   - score <= 1 OR CI red        → auto bounce with feedback card
+
+Step 4's bounce hands off to :mod:`app.services.revision`.
+"""
+
+from __future__ import annotations
+
+import asyncio
+import json
+import logging
+import re
+from dataclasses import asdict, dataclass, field
+from typing import Any, Optional
+
+from app.config import settings
+from app.services.acceptance import AcceptanceContract
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class StaticIssue:
+    rule: str           # max_pr_lines / forbid_files / coverage_min / ci_required
+    message: str
+    severity: str = "blocking"  # blocking | warning
+
+
+@dataclass
+class LLMIssue:
+    category: str       # responds_to_issue / antipattern / commit_quality / other
+    message: str
+    file: Optional[str] = None
+    line: Optional[int] = None
+    how_to_fix: str = ""  # required field — must always be present
+    severity: str = "blocking"
+
+
+@dataclass
+class ReviewResult:
+    score: int                                 # 1-5
+    decision: str                              # auto_merge | hr_review | auto_bounce
+    static_issues: list[StaticIssue] = field(default_factory=list)
+    llm_issues: list[LLMIssue] = field(default_factory=list)
+    ci_passed: bool = True
+    notes: str = ""
+
+    def to_dict(self) -> dict[str, Any]:
+        return asdict(self)
+
+
+def evaluate_static_rules(
+    diff_stats: dict[str, Any],
+    contract: AcceptanceContract,
+    *,
+    ci_results: Optional[dict[str, str]] = None,
+    coverage_pct: Optional[float] = None,
+) -> list[StaticIssue]:
+    issues: list[StaticIssue] = []
+
+    # max_pr_lines
+    if contract.max_pr_lines:
+        total = int(diff_stats.get("additions", 0)) + int(diff_stats.get("deletions", 0))
+        if total > contract.max_pr_lines:
+            issues.append(StaticIssue(
+                rule="max_pr_lines",
+                message=f"PR {total} 行 > max_pr_lines={contract.max_pr_lines},请拆分。",
+            ))
+
+    # forbid_files
+    changed_files = diff_stats.get("changed_files") or []
+    for pattern in contract.forbid_files:
+        for f in changed_files:
+            if _glob_match(pattern, f):
+                issues.append(StaticIssue(
+                    rule="forbid_files",
+                    message=f"禁止修改 `{f}` (匹配 {pattern})",
+                ))
+
+    # coverage_min
+    if contract.coverage_min and coverage_pct is not None and coverage_pct < contract.coverage_min:
+        issues.append(StaticIssue(
+            rule="coverage_min",
+            message=f"覆盖率 {coverage_pct:.1f}% < {contract.coverage_min}%",
+        ))
+
+    # ci_required missing/failed
+    if contract.ci_required and ci_results is not None:
+        for required in contract.ci_required:
+            status = ci_results.get(required)
+            if status != "success":
+                issues.append(StaticIssue(
+                    rule="ci_required",
+                    message=f"CI job `{required}` 状态: {status or 'missing'}",
+                ))
+
+    return issues
+
+
+def _glob_match(pattern: str, path: str) -> bool:
+    """Tiny glob — supports `*` (filename only) and exact path match."""
+    import fnmatch
+    return fnmatch.fnmatch(path, pattern) or path == pattern
+
+
+_LLM_REVIEW_SYSTEM = """你是代码 PR 审查助手,根据任务 issue 描述判断 PR 是否真的解决了问题。
+评分标准(1-5,5 最好):
+- 5: 完美解决 + 测试到位 + 无反模式
+- 4: 解决了主要问题,小瑕疵
+- 3: 大致解决但有疑虑
+- 2: 走捷径或绕过关键点
+- 1: 没解决,甚至更糟
+
+每个 issue 必须包含 `how_to_fix` 字段(具体修法),缺失则视为无效。
+
+返回 JSON:
+{
+  "score": 4,
+  "issues": [
+    {"category": "antipattern", "message": "...", "file": "x.py", "line": 12,
+     "how_to_fix": "用 select_related 替代 N+1", "severity": "blocking"}
+  ],
+  "summary": "简评一句"
+}
+只返回 JSON。
+"""
+
+
+async def llm_review(
+    task_data: dict[str, Any],
+    diff_text: str,
+    *,
+    llm: Any = None,
+    timeout: int = 30,
+) -> tuple[int, list[LLMIssue], str]:
+    """Score 1-5 + structured issues. Retries once if how_to_fix missing."""
+    if llm is None:
+        from app.services.llm import llm_service
+        llm = llm_service
+
+    prompt = _build_llm_prompt(task_data, diff_text)
+    try:
+        raw = await asyncio.wait_for(llm.call(prompt, _LLM_REVIEW_SYSTEM), timeout=timeout)
+    except Exception as exc:
+        logger.warning("llm_review: call failed: %s; defaulting to score=3", exc)
+        return 3, [], "LLM 不可用,降级人工审"
+
+    parsed = _parse_review_json(raw)
+    if parsed is None:
+        return 3, [], "LLM 响应解析失败,降级人工审"
+
+    score = int(parsed.get("score", 3))
+    issues = [_to_llm_issue(i) for i in parsed.get("issues", [])]
+    if any(not i.how_to_fix for i in issues):
+        # Retry once: ask LLM to fix missing how_to_fix
+        try:
+            raw2 = await asyncio.wait_for(
+                llm.call(prompt + "\n\n注意:每条 issue 必须有 how_to_fix 字段。",
+                         _LLM_REVIEW_SYSTEM),
+                timeout=timeout,
+            )
+            parsed2 = _parse_review_json(raw2)
+            if parsed2 is not None:
+                issues2 = [_to_llm_issue(i) for i in parsed2.get("issues", [])]
+                if all(i.how_to_fix for i in issues2):
+                    issues = issues2
+        except Exception:
+            pass
+
+    # Final degradation: stamp a templated how_to_fix on any still-missing
+    for issue in issues:
+        if not issue.how_to_fix:
+            issue.how_to_fix = "(LLM 未给具体修法,请人工评审后补充)"
+
+    return max(1, min(5, score)), issues, str(parsed.get("summary", ""))
+
+
+def _build_llm_prompt(task: dict[str, Any], diff: str) -> str:
+    return (
+        f"任务标题: {task.get('title', '')}\n"
+        f"任务描述: {task.get('description', '')}\n\n"
+        f"PR diff:\n```diff\n{diff[:8000]}\n```\n"
+    )
+
+
+def _parse_review_json(raw: str) -> Optional[dict]:
+    if not raw:
+        return None
+    fenced = re.search(r"```(?:json)?\s*(\{.*?\})\s*```", raw, re.DOTALL)
+    payload = fenced.group(1) if fenced else None
+    if payload is None:
+        obj = re.search(r"\{.*\}", raw, re.DOTALL)
+        payload = obj.group(0) if obj else None
+    if not payload:
+        return None
+    try:
+        return json.loads(payload)
+    except json.JSONDecodeError:
+        return None
+
+
+def _to_llm_issue(d: dict[str, Any]) -> LLMIssue:
+    return LLMIssue(
+        category=str(d.get("category", "other")),
+        message=str(d.get("message", "")).strip(),
+        file=d.get("file"),
+        line=d.get("line"),
+        how_to_fix=str(d.get("how_to_fix", "")).strip(),
+        severity=str(d.get("severity", "blocking")),
+    )
+
+
+def decide(
+    score: int,
+    static_issues: list[StaticIssue],
+    ci_passed: bool,
+    threshold: Optional[int] = None,
+) -> str:
+    """Return one of: auto_merge | hr_review | auto_bounce."""
+    threshold = threshold or settings.auto_verify_threshold
+    blocking_static = [s for s in static_issues if s.severity == "blocking"]
+    if not ci_passed or blocking_static or score <= 1:
+        return "auto_bounce"
+    if score >= threshold:
+        return "auto_merge"
+    return "hr_review"

--- a/app/services/proactive.py
+++ b/app/services/proactive.py
@@ -1,0 +1,285 @@
+"""Proactive task tracking — issue #12 §1, §2.
+
+Scans IN_PROGRESS tasks every PROACTIVE_SCAN_INTERVAL_HOURS (default 6h).
+Emits one of four signals per task based on time-to-deadline:
+
+| Trigger band                  | Action                                |
+| ----------------------------- | ------------------------------------- |
+| 50% of total window, idle     | Gentle reminder card → candidate      |
+| 24h before deadline           | Escalation card → candidate + HR      |
+| 2h before deadline            | Urgent card → candidate + HR          |
+| Past deadline, undelivered    | Mark OVERDUE + risk card → HR         |
+
+A task with a progress update in the last PROGRESS_IDLE_GRACE_HOURS
+(default 24h) skips the 50% gentle reminder for that scan.
+"""
+
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from datetime import datetime, timedelta, timezone, UTC
+from enum import Enum
+from typing import Any, Optional
+
+from app.config import settings
+
+logger = logging.getLogger(__name__)
+
+
+class ReminderTier(str, Enum):
+    GENTLE_50 = "gentle_50"
+    ESCALATE_24H = "escalate_24h"
+    URGENT_2H = "urgent_2h"
+    OVERDUE = "overdue"
+    SKIP = "skip"
+
+
+@dataclass
+class ScanDecision:
+    task_id: str
+    tier: ReminderTier
+    reason: str = ""
+
+
+def classify_task(task: dict[str, Any], now: Optional[datetime] = None) -> ScanDecision:
+    """Pure classifier — given a task row + current time, return the tier."""
+    now = now or datetime.now(UTC)
+    task_id = task.get("task_id") or task.get("taskid") or ""
+
+    # Skip non-active tasks
+    if task.get("status") not in {"IN_PROGRESS", "in_progress"}:
+        return ScanDecision(task_id, ReminderTier.SKIP, "not in progress")
+
+    deadline_str = task.get("deadline")
+    started_str = task.get("started_at") or task.get("accepted_at") or task.get("created_at")
+    if not deadline_str or not started_str:
+        return ScanDecision(task_id, ReminderTier.SKIP, "missing timestamps")
+
+    try:
+        deadline = _parse_dt(deadline_str)
+        started = _parse_dt(started_str)
+    except ValueError:
+        return ScanDecision(task_id, ReminderTier.SKIP, "unparseable timestamps")
+
+    # Past deadline
+    if now >= deadline:
+        return ScanDecision(task_id, ReminderTier.OVERDUE, "past deadline")
+
+    time_left = deadline - now
+    # Urgent: ≤ 2h to deadline
+    if time_left <= timedelta(hours=2):
+        return ScanDecision(task_id, ReminderTier.URGENT_2H, "≤ 2h remaining")
+    # Escalate: ≤ 24h to deadline
+    if time_left <= timedelta(hours=24):
+        return ScanDecision(task_id, ReminderTier.ESCALATE_24H, "≤ 24h remaining")
+
+    # 50% gentle band — only if idle (no progress in last grace window)
+    total = deadline - started
+    if total.total_seconds() <= 0:
+        return ScanDecision(task_id, ReminderTier.SKIP, "non-positive window")
+    elapsed_ratio = (now - started) / total
+    if elapsed_ratio >= 0.5 and not _has_recent_progress(task, now):
+        return ScanDecision(task_id, ReminderTier.GENTLE_50, "50% window + idle")
+
+    return ScanDecision(task_id, ReminderTier.SKIP, "within healthy window")
+
+
+def _has_recent_progress(task: dict[str, Any], now: datetime) -> bool:
+    """Check progress_updates for an entry within PROGRESS_IDLE_GRACE_HOURS."""
+    updates = task.get("progress_updates") or []
+    if not updates:
+        last = task.get("last_progress_at")
+        if not last:
+            return False
+        try:
+            last_dt = _parse_dt(last)
+        except ValueError:
+            return False
+        return (now - last_dt) <= timedelta(hours=settings.progress_idle_grace_hours)
+
+    latest = updates[-1] if isinstance(updates, list) else None
+    if not isinstance(latest, dict) or "timestamp" not in latest:
+        return False
+    try:
+        latest_dt = _parse_dt(latest["timestamp"])
+    except ValueError:
+        return False
+    return (now - latest_dt) <= timedelta(hours=settings.progress_idle_grace_hours)
+
+
+def _parse_dt(value: str) -> datetime:
+    s = str(value).replace("Z", "+00:00")
+    # Plain date "2026-12-01" → end-of-day UTC for fairness.
+    if len(s) == 10 and s.count("-") == 2:
+        s += "T23:59:59+00:00"
+    dt = datetime.fromisoformat(s)
+    if dt.tzinfo is None:
+        dt = dt.replace(tzinfo=UTC)
+    return dt
+
+
+# ----------------------------------------------------------------------
+# Service: drives the scan and dispatches to candidate / HR
+# ----------------------------------------------------------------------
+
+class ProactiveTracker:
+    def __init__(
+        self,
+        task_manager: Any = None,
+        sender: Any = None,
+    ) -> None:
+        if task_manager is None:
+            from app.services.task_manager import task_manager as _tm
+            task_manager = _tm
+        if sender is None:
+            from app.services.feishu_im import FeishuIMSender
+            sender = FeishuIMSender()
+        self.tm = task_manager
+        self.sender = sender
+
+    async def scan_once(
+        self, now: Optional[datetime] = None
+    ) -> list[ScanDecision]:
+        """Run one scan pass; returns the list of decisions made."""
+        now = now or datetime.now(UTC)
+        try:
+            tasks = await self.tm.bitable.list_in_progress_tasks()
+        except AttributeError:
+            tasks = await self._fallback_list(now)
+        except Exception as exc:
+            logger.warning("ProactiveTracker.scan_once: list failed: %s", exc)
+            return []
+
+        decisions: list[ScanDecision] = []
+        for task in tasks or []:
+            decision = classify_task(task, now=now)
+            if decision.tier is ReminderTier.SKIP:
+                continue
+            await self._dispatch(task, decision)
+            decisions.append(decision)
+        return decisions
+
+    async def _dispatch(self, task: dict[str, Any], decision: ScanDecision) -> None:
+        candidate = task.get("assignee") or task.get("current_candidate")
+        creator = task.get("created_by") or task.get("creator")
+        title = task.get("title", "(任务)")
+
+        if decision.tier is ReminderTier.GENTLE_50:
+            if candidate:
+                await self.sender.push_text(
+                    candidate, f"⏰ 任务 {title} 已过半,记得用 /progress 同步进度。"
+                )
+        elif decision.tier is ReminderTier.ESCALATE_24H:
+            if candidate:
+                await self.sender.push_text(
+                    candidate, f"⚠️ 任务 {title} DDL 不足 24h,请尽快交付或同步风险。"
+                )
+            if creator:
+                await self.sender.push_text(
+                    creator, f"⚠️ 任务 {title} 24h 内将到期,候选人 {candidate} 暂未交付。"
+                )
+        elif decision.tier is ReminderTier.URGENT_2H:
+            if candidate:
+                await self.sender.push_text(
+                    candidate, f"🚨 任务 {title} 距 DDL 不足 2h,请立即响应。"
+                )
+            if creator:
+                await self.sender.push_text(
+                    creator, f"🚨 任务 {title} 2h 后到期,需关注。"
+                )
+        elif decision.tier is ReminderTier.OVERDUE:
+            try:
+                await self.tm.bitable.update_task(decision.task_id, {"status": "OVERDUE"})
+            except Exception as exc:  # pragma: no cover
+                logger.warning("OVERDUE status update failed: %s", exc)
+            if creator:
+                await self.sender.push_text(
+                    creator,
+                    f"❌ 任务 {title} 已逾期未交付。可改派或延期,请处理。",
+                )
+
+    async def _fallback_list(self, now: datetime) -> list[dict[str, Any]]:
+        try:
+            data = await self.tm.bitable.get_all_tasks_sorted(page_size=200, page=0)
+        except Exception:
+            return []
+        items = data.get("items") if isinstance(data, dict) else data
+        return [
+            t for t in (items or [])
+            if (t.get("status") or "").lower() == "in_progress"
+        ]
+
+
+# ----------------------------------------------------------------------
+# Progress command handler — extracts percentage / risk via LLM
+# ----------------------------------------------------------------------
+
+async def record_progress(
+    task_id: str,
+    text: str,
+    *,
+    task_manager: Any = None,
+    llm: Any = None,
+) -> dict[str, Any]:
+    """Append a progress update with LLM-extracted percentage + risks.
+
+    Stored on the task row as ``progress_updates`` (JSON array of
+    ``{timestamp, text, percentage, risks}``). Last entry's timestamp
+    is used by the proactive scan to skip the 50% gentle reminder.
+    """
+    if task_manager is None:
+        from app.services.task_manager import task_manager as _tm
+        task_manager = _tm
+    if llm is None:
+        from app.services.llm import llm_service
+        llm = llm_service
+
+    percentage, risks = await _extract_progress(llm, text)
+    entry = {
+        "timestamp": datetime.now(UTC).isoformat(),
+        "text": text,
+        "percentage": percentage,
+        "risks": risks,
+    }
+    try:
+        task = await task_manager.bitable.get_task(task_id) or {}
+        history = list(task.get("progress_updates") or [])
+        history.append(entry)
+        await task_manager.bitable.update_task(
+            task_id, {"progress_updates": history, "last_progress_at": entry["timestamp"]}
+        )
+    except Exception as exc:
+        logger.warning("record_progress: persist failed: %s", exc)
+    return entry
+
+
+async def _extract_progress(llm: Any, text: str) -> tuple[Optional[int], list[str]]:
+    """Best-effort: ask LLM for {percentage:int, risks:[str]}; tolerate failure."""
+    if not text or llm is None:
+        return None, []
+    try:
+        import asyncio
+        import json
+        import re
+
+        prompt = (
+            "从以下进度报告中抽取百分比(0-100)和风险点(列表)。"
+            "返回 JSON: {\"percentage\":N, \"risks\":[\"...\"]}\n\n"
+            f"报告: {text}"
+        )
+        response = await asyncio.wait_for(llm.call(prompt, ""), timeout=10)
+        match = re.search(r"\{.*\}", response, re.DOTALL)
+        if not match:
+            return None, []
+        data = json.loads(match.group(0))
+        pct = data.get("percentage")
+        if isinstance(pct, str) and pct.isdigit():
+            pct = int(pct)
+        risks = data.get("risks") or []
+        if not isinstance(risks, list):
+            risks = [str(risks)]
+        return (pct if isinstance(pct, int) else None), [str(r) for r in risks]
+    except Exception as exc:
+        logger.debug("progress extraction failed: %s", exc)
+        return None, []

--- a/app/services/revision.py
+++ b/app/services/revision.py
@@ -1,0 +1,159 @@
+"""Revision loop — issue #12 §5.
+
+Builds the structured feedback card pushed to a candidate when a PR is
+bounced (CI red, LLM low score, static rule violation). Tracks
+``revision_count`` and ``revision_history`` on the task row, escalates
+to HR when MAX_REVISIONS reached.
+"""
+
+from __future__ import annotations
+
+import logging
+from datetime import UTC, datetime
+from typing import Any, Optional
+
+from app.config import settings
+from app.services.ai_review import LLMIssue, StaticIssue
+
+logger = logging.getLogger(__name__)
+
+
+def build_feedback_card(
+    task: dict[str, Any],
+    *,
+    ci_failures: list[str],
+    static_issues: list[StaticIssue],
+    llm_issues: list[LLMIssue],
+    style_tips: Optional[list[str]] = None,
+    revision_count: int = 0,
+) -> dict[str, Any]:
+    """Card with 4 sections: CI / LLM / static rules / style tips.
+
+    Each section is omitted when empty so the candidate isn't shown
+    "(none)" placeholders. Issues with file/line anchors render as
+    ``file:line — message``; ``how_to_fix`` is always shown.
+    """
+    elements = [
+        {"tag": "markdown",
+         "content": f"## 修订反馈 — 第 {revision_count + 1} 轮 / 上限 {settings.max_revisions}"}
+    ]
+
+    if ci_failures:
+        elements.append({"tag": "markdown", "content": "**CI 失败摘要**"})
+        for line in ci_failures:
+            elements.append({"tag": "markdown", "content": f"- {line}"})
+
+    if llm_issues:
+        elements.append({"tag": "markdown", "content": "**LLM 评注**"})
+        for issue in llm_issues:
+            anchor = f"`{issue.file}:{issue.line}` " if issue.file and issue.line else ""
+            elements.append({
+                "tag": "markdown",
+                "content": f"- {anchor}{issue.message}\n  - 修法: {issue.how_to_fix}",
+            })
+
+    blocking_static = [s for s in static_issues if s.severity != "warning"]
+    if blocking_static:
+        elements.append({"tag": "markdown", "content": "**静态规则违规**"})
+        for s in blocking_static:
+            elements.append({"tag": "markdown", "content": f"- ({s.rule}) {s.message}"})
+
+    style = list(style_tips or []) + [
+        s.message for s in static_issues if s.severity == "warning"
+    ]
+    if style:
+        elements.append({"tag": "markdown", "content": "**风格提示(不阻塞)**"})
+        for s in style:
+            elements.append({"tag": "markdown", "content": f"- {s}"})
+
+    elements.append({
+        "tag": "markdown",
+        "content": "💡 push 新 commit 自动触发重检。",
+    })
+
+    return {
+        "schema": "2.0",
+        "header": {
+            "title": {"tag": "plain_text", "content": "修订反馈"},
+            "template": "red",
+        },
+        "body": {"elements": elements},
+    }
+
+
+async def record_revision(
+    task_id: str,
+    *,
+    commit_sha: str,
+    issues_summary: dict[str, Any],
+    task_manager: Any = None,
+) -> dict[str, Any]:
+    """Append to revision_history, increment revision_count, persist.
+
+    Returns a dict with the new count and a flag ``escalate=True`` if the
+    cap was hit (caller must page HR + show full history).
+    """
+    if task_manager is None:
+        from app.services.task_manager import task_manager as _tm
+        task_manager = _tm
+
+    try:
+        task = (await task_manager.bitable.get_task(task_id)) or {}
+    except Exception as exc:
+        logger.warning("record_revision: get_task failed: %s", exc)
+        task = {}
+
+    history = list(task.get("revision_history") or [])
+    history.append({
+        "round": len(history) + 1,
+        "commit_sha": commit_sha,
+        "issues": issues_summary,
+        "timestamp": datetime.now(UTC).isoformat(),
+    })
+    new_count = len(history)
+    escalate = new_count >= settings.max_revisions
+
+    update_payload = {
+        "revision_count": new_count,
+        "revision_history": history,
+    }
+    if escalate:
+        update_payload["status"] = "ESCALATED"
+
+    try:
+        await task_manager.bitable.update_task(task_id, update_payload)
+    except Exception as exc:
+        logger.warning("record_revision: update_task failed: %s", exc)
+
+    return {
+        "revision_count": new_count,
+        "history": history,
+        "escalate": escalate,
+    }
+
+
+def build_escalation_card(
+    task: dict[str, Any], history: list[dict[str, Any]]
+) -> dict[str, Any]:
+    """Card for HR when MAX_REVISIONS is hit; shows full revision_history."""
+    rounds = [
+        f"- 第 {h.get('round')} 轮 (commit {h.get('commit_sha', '')[:8]}) — "
+        f"{(h.get('issues') or {}).get('summary', '(无摘要)')}"
+        for h in history
+    ]
+    return {
+        "schema": "2.0",
+        "header": {
+            "title": {"tag": "plain_text",
+                      "content": f"任务升级 — {task.get('title', '')}"},
+            "template": "purple",
+        },
+        "body": {
+            "elements": [
+                {"tag": "markdown",
+                 "content": f"修订已达 {len(history)} 轮(上限 {settings.max_revisions}),请人工介入。"},
+                {"tag": "markdown", "content": "**修订历史**"},
+                {"tag": "markdown", "content": "\n".join(rounds) or "_(空)_"},
+            ]
+        },
+    }

--- a/tests/integration/test_ai_verification.py
+++ b/tests/integration/test_ai_verification.py
@@ -1,0 +1,150 @@
+"""Integration tests for AI pre-verification (issue #12 §3, §4)."""
+
+from __future__ import annotations
+
+import json
+from unittest.mock import AsyncMock, MagicMock
+
+import pytest
+
+from app.services.acceptance import AcceptanceContract, parse_acceptance_yaml
+from app.services.ai_review import (
+    LLMIssue,
+    StaticIssue,
+    decide,
+    evaluate_static_rules,
+    llm_review,
+)
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 1
+# ----------------------------------------------------------------------
+@pytest.mark.asyncio
+async def test_green_ci_high_score_auto_merge():
+    """Green CI + LLM score ≥ threshold → decide returns auto_merge."""
+    llm = MagicMock()
+    llm.call = AsyncMock(return_value=json.dumps({
+        "score": 5,
+        "issues": [],
+        "summary": "Clean fix.",
+    }))
+    score, issues, _ = await llm_review(
+        {"title": "X", "description": "fix bug"}, "diff text", llm=llm,
+    )
+    assert score == 5
+    assert issues == []
+    decision = decide(score, [], ci_passed=True)
+    assert decision == "auto_merge"
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 2
+# ----------------------------------------------------------------------
+def test_low_score_routes_to_hr():
+    """2 ≤ score < threshold AND CI green → hr_review."""
+    assert decide(3, [], ci_passed=True, threshold=4) == "hr_review"
+    assert decide(2, [], ci_passed=True, threshold=4) == "hr_review"
+
+
+def test_red_ci_routes_to_auto_bounce_regardless_of_score():
+    """CI red → auto_bounce even if LLM score is high."""
+    assert decide(5, [], ci_passed=False) == "auto_bounce"
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 3
+# ----------------------------------------------------------------------
+def test_acceptance_yaml_parsed_from_issue():
+    """Issue body with the documented YAML fence parses into a contract."""
+    body = """
+Some prose about the task.
+
+Acceptance criteria:
+
+```yaml
+acceptance:
+  ci_required:
+    - test
+    - lint
+  coverage_min: 80
+  custom_commands:
+    - "pytest tests/integration/test_xxx.py"
+  forbid_files:
+    - .env
+    - "*.pem"
+  max_pr_lines: 2000
+```
+
+More prose.
+"""
+    contract = parse_acceptance_yaml(body)
+    assert contract.ci_required == ["test", "lint"]
+    assert contract.coverage_min == 80
+    assert contract.max_pr_lines == 2000
+    assert ".env" in contract.forbid_files
+    assert "*.pem" in contract.forbid_files
+    assert "pytest tests/integration/test_xxx.py" in contract.custom_commands
+
+
+def test_acceptance_missing_yaml_returns_default():
+    contract = parse_acceptance_yaml("body without any yaml block")
+    assert contract.is_default
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 4
+# ----------------------------------------------------------------------
+def test_max_pr_lines_violation_auto_bounce():
+    """PR exceeds max_pr_lines → static issue + auto_bounce."""
+    contract = AcceptanceContract(max_pr_lines=2000)
+    diff_stats = {"additions": 1500, "deletions": 700}
+    static = evaluate_static_rules(diff_stats, contract)
+    assert any(s.rule == "max_pr_lines" for s in static)
+    decision = decide(score=5, static_issues=static, ci_passed=True)
+    assert decision == "auto_bounce"
+
+
+def test_forbid_files_static_check():
+    """Touching a forbidden file → blocking static issue."""
+    contract = AcceptanceContract(forbid_files=[".env", "*.pem"])
+    diff_stats = {"additions": 10, "deletions": 0,
+                  "changed_files": [".env", "src/main.py"]}
+    static = evaluate_static_rules(diff_stats, contract)
+    assert any(s.rule == "forbid_files" for s in static)
+
+
+def test_coverage_min_static_check():
+    contract = AcceptanceContract(coverage_min=80)
+    diff_stats = {"additions": 1, "deletions": 1}
+    static = evaluate_static_rules(diff_stats, contract, coverage_pct=72.5)
+    assert any(s.rule == "coverage_min" for s in static)
+
+
+@pytest.mark.asyncio
+async def test_llm_review_retries_on_missing_how_to_fix():
+    """First response missing how_to_fix → retry; final fallback fills template."""
+    llm = MagicMock()
+    bad = json.dumps({"score": 3, "issues": [
+        {"category": "antipattern", "message": "N+1", "file": "x.py", "line": 5},
+    ]})
+    good = json.dumps({"score": 3, "issues": [
+        {"category": "antipattern", "message": "N+1", "file": "x.py", "line": 5,
+         "how_to_fix": "Use select_related()"},
+    ]})
+    llm.call = AsyncMock(side_effect=[bad, good])
+    score, issues, _ = await llm_review({"title": "T", "description": ""}, "diff", llm=llm)
+    assert llm.call.await_count == 2
+    assert all(i.how_to_fix for i in issues)
+
+
+@pytest.mark.asyncio
+async def test_llm_review_template_fallback_when_retry_also_missing():
+    """If retry STILL omits how_to_fix, we stamp a templated value."""
+    llm = MagicMock()
+    bad = json.dumps({"score": 2, "issues": [
+        {"category": "other", "message": "msg", "file": None, "line": None},
+    ]})
+    llm.call = AsyncMock(return_value=bad)
+    _, issues, _ = await llm_review({"title": "T", "description": ""}, "diff", llm=llm)
+    assert all(i.how_to_fix for i in issues)

--- a/tests/integration/test_proactive_tracking.py
+++ b/tests/integration/test_proactive_tracking.py
@@ -1,0 +1,130 @@
+"""Integration tests for the proactive tracking scan (issue #12 §1, §2)."""
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta, timezone, UTC
+from unittest.mock import AsyncMock, MagicMock
+
+import pytest
+
+from app.services.proactive import (
+    ProactiveTracker,
+    ReminderTier,
+    classify_task,
+    record_progress,
+)
+
+
+def _task(**overrides):
+    base = {
+        "task_id": "T1",
+        "taskid": "T1",
+        "title": "Task",
+        "status": "IN_PROGRESS",
+        "deadline": "2026-12-01T00:00:00+00:00",
+        "started_at": "2026-11-01T00:00:00+00:00",
+        "assignee": "user_a",
+        "created_by": "hr_001",
+        "progress_updates": [],
+    }
+    base.update(overrides)
+    return base
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 1
+# ----------------------------------------------------------------------
+def test_idle_50pct_ddl_triggers_reminder():
+    """Half-way through window with no recent progress → GENTLE_50."""
+    started = datetime(2026, 1, 1, tzinfo=UTC)
+    deadline = started + timedelta(days=10)
+    now = started + timedelta(days=6)  # 60% in
+    decision = classify_task(_task(
+        started_at=started.isoformat(),
+        deadline=deadline.isoformat(),
+        progress_updates=[],
+    ), now=now)
+    assert decision.tier is ReminderTier.GENTLE_50
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 2
+# ----------------------------------------------------------------------
+def test_ddl_24h_escalates_to_hr():
+    started = datetime(2026, 1, 1, tzinfo=UTC)
+    deadline = started + timedelta(days=10)
+    now = deadline - timedelta(hours=20)
+    decision = classify_task(_task(
+        started_at=started.isoformat(),
+        deadline=deadline.isoformat(),
+    ), now=now)
+    assert decision.tier is ReminderTier.ESCALATE_24H
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 3
+# ----------------------------------------------------------------------
+def test_progress_command_resets_idle_timer():
+    """Recent progress update suppresses the GENTLE_50 reminder."""
+    started = datetime(2026, 1, 1, tzinfo=UTC)
+    deadline = started + timedelta(days=10)
+    now = started + timedelta(days=6)
+    recent_update = {"timestamp": (now - timedelta(hours=2)).isoformat(),
+                     "text": "70% done"}
+    decision = classify_task(_task(
+        started_at=started.isoformat(),
+        deadline=deadline.isoformat(),
+        progress_updates=[recent_update],
+    ), now=now)
+    assert decision.tier is ReminderTier.SKIP
+
+
+@pytest.mark.asyncio
+async def test_record_progress_appends_update():
+    tm = MagicMock()
+    tm.bitable = MagicMock()
+    tm.bitable.get_task = AsyncMock(return_value=_task(progress_updates=[]))
+    tm.bitable.update_task = AsyncMock(return_value=True)
+
+    llm = MagicMock()
+    llm.call = AsyncMock(return_value='{"percentage": 60, "risks": ["数据源缺字段"]}')
+
+    entry = await record_progress("T1", "已完成数据拉取,正在处理", task_manager=tm, llm=llm)
+
+    assert entry["percentage"] == 60
+    assert entry["risks"] == ["数据源缺字段"]
+    update = tm.bitable.update_task.await_args.args[1]
+    assert "progress_updates" in update
+    assert update["progress_updates"][-1]["text"] == "已完成数据拉取,正在处理"
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 4
+# ----------------------------------------------------------------------
+@pytest.mark.asyncio
+async def test_overdue_marks_status_and_notifies_hr():
+    """Past deadline → tier OVERDUE → status=OVERDUE + HR notification."""
+    started = datetime(2026, 1, 1, tzinfo=UTC)
+    deadline = datetime(2026, 1, 5, tzinfo=UTC)
+    now = datetime(2026, 1, 6, tzinfo=UTC)
+
+    sender = MagicMock()
+    sender.push_text = AsyncMock()
+    sender.push_card = AsyncMock()
+    tm = MagicMock()
+    tm.bitable = MagicMock()
+    tm.bitable.list_in_progress_tasks = AsyncMock(return_value=[
+        _task(started_at=started.isoformat(), deadline=deadline.isoformat()),
+    ])
+    tm.bitable.update_task = AsyncMock(return_value=True)
+
+    tracker = ProactiveTracker(task_manager=tm, sender=sender)
+    decisions = await tracker.scan_once(now=now)
+
+    assert any(d.tier is ReminderTier.OVERDUE for d in decisions)
+    update = tm.bitable.update_task.await_args.args[1]
+    assert update.get("status") == "OVERDUE"
+    sender.push_text.assert_awaited()
+    text = sender.push_text.await_args_list[0]
+    assert text.args[0] == "hr_001"
+    assert "逾期" in text.args[1]

--- a/tests/integration/test_revision_loop.py
+++ b/tests/integration/test_revision_loop.py
@@ -1,0 +1,186 @@
+"""Integration tests for the revision loop (issue #12 §5, §6)."""
+
+from __future__ import annotations
+
+import asyncio
+from unittest.mock import AsyncMock, MagicMock, patch
+
+import pytest
+
+from app.services.ai_review import LLMIssue, StaticIssue
+from app.services.revision import (
+    build_escalation_card,
+    build_feedback_card,
+    record_revision,
+)
+
+
+def _task(**overrides):
+    base = {
+        "task_id": "T1",
+        "taskid": "T1",
+        "title": "Sample",
+        "revision_count": 0,
+        "revision_history": [],
+        "assignee": "user_a",
+        "created_by": "hr_001",
+    }
+    base.update(overrides)
+    return base
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 1
+# ----------------------------------------------------------------------
+def test_ci_fail_pushes_actionable_card():
+    """CI failure section appears in the feedback card."""
+    card = build_feedback_card(
+        _task(),
+        ci_failures=["pytest tests/test_foo.py FAILED — AssertionError: expected 0, got 3"],
+        static_issues=[],
+        llm_issues=[],
+    )
+    elements = card["body"]["elements"]
+    found = [e for e in elements if "CI 失败" in e.get("content", "")]
+    assert found, "Feedback card must include a CI failures section"
+    detail = next(e for e in elements if "expected 0" in e.get("content", ""))
+    assert detail
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 2
+# ----------------------------------------------------------------------
+def test_feedback_includes_file_line_anchors():
+    """LLM issues with file/line render as `file:line — message`."""
+    card = build_feedback_card(
+        _task(),
+        ci_failures=[],
+        static_issues=[],
+        llm_issues=[
+            LLMIssue(
+                category="antipattern",
+                message="N+1 query in for-loop",
+                file="app/views.py",
+                line=42,
+                how_to_fix="Use prefetch_related('items')",
+            )
+        ],
+    )
+    text = "\n".join(e.get("content", "") for e in card["body"]["elements"])
+    assert "app/views.py:42" in text
+    assert "N+1 query in for-loop" in text
+    assert "prefetch_related" in text
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 3
+# ----------------------------------------------------------------------
+@pytest.mark.asyncio
+async def test_new_commit_triggers_recheck():
+    """A new commit_sha → record_revision appends a new round."""
+    tm = MagicMock()
+    tm.bitable = MagicMock()
+    tm.bitable.get_task = AsyncMock(return_value=_task(revision_history=[
+        {"round": 1, "commit_sha": "aaa111", "issues": {"summary": "first"},
+         "timestamp": "x"}
+    ]))
+    tm.bitable.update_task = AsyncMock(return_value=True)
+
+    result = await record_revision(
+        "T1", commit_sha="bbb222",
+        issues_summary={"summary": "second round"},
+        task_manager=tm,
+    )
+    assert result["revision_count"] == 2
+    payload = tm.bitable.update_task.await_args.args[1]
+    assert payload["revision_count"] == 2
+    assert payload["revision_history"][-1]["commit_sha"] == "bbb222"
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 4
+# ----------------------------------------------------------------------
+@pytest.mark.asyncio
+async def test_revision_count_increments():
+    tm = MagicMock()
+    tm.bitable = MagicMock()
+    tm.bitable.get_task = AsyncMock(return_value=_task(revision_history=[]))
+    tm.bitable.update_task = AsyncMock(return_value=True)
+
+    r1 = await record_revision("T1", commit_sha="c1",
+                               issues_summary={"summary": "a"}, task_manager=tm)
+    assert r1["revision_count"] == 1
+    # Simulate persisted history for round 2
+    tm.bitable.get_task = AsyncMock(return_value=_task(revision_history=r1["history"]))
+    r2 = await record_revision("T1", commit_sha="c2",
+                               issues_summary={"summary": "b"}, task_manager=tm)
+    assert r2["revision_count"] == 2
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 5
+# ----------------------------------------------------------------------
+@pytest.mark.asyncio
+async def test_max_revisions_escalates_with_history():
+    """Reaching MAX_REVISIONS sets status=ESCALATED and escalate flag."""
+    from app.config import settings
+    history = [
+        {"round": i + 1, "commit_sha": f"sha{i}", "issues": {"summary": f"r{i}"},
+         "timestamp": "x"}
+        for i in range(settings.max_revisions - 1)
+    ]
+    tm = MagicMock()
+    tm.bitable = MagicMock()
+    tm.bitable.get_task = AsyncMock(return_value=_task(revision_history=history))
+    tm.bitable.update_task = AsyncMock(return_value=True)
+
+    result = await record_revision(
+        "T1", commit_sha="finalsha",
+        issues_summary={"summary": "final"}, task_manager=tm,
+    )
+    assert result["escalate"] is True
+    assert result["revision_count"] == settings.max_revisions
+    payload = tm.bitable.update_task.await_args.args[1]
+    assert payload["status"] == "ESCALATED"
+
+    # Escalation card displays the full history.
+    card = build_escalation_card(_task(), result["history"])
+    text = "\n".join(e.get("content", "") for e in card["body"]["elements"])
+    assert all(f"sha{i}"[:8] in text for i in range(settings.max_revisions - 1))
+
+
+# ----------------------------------------------------------------------
+# Acceptance case 6
+# ----------------------------------------------------------------------
+def test_how_to_fix_field_always_present():
+    """Every LLMIssue rendered into the card has a how_to_fix line."""
+    card = build_feedback_card(
+        _task(),
+        ci_failures=[],
+        static_issues=[],
+        llm_issues=[
+            LLMIssue(category="other", message="m1", how_to_fix="hf-1"),
+            LLMIssue(category="other", message="m2", how_to_fix="hf-2"),
+        ],
+    )
+    text = "\n".join(e.get("content", "") for e in card["body"]["elements"])
+    # Each how_to_fix appears as "修法: ..."
+    assert text.count("修法: ") == 2
+    assert "hf-1" in text and "hf-2" in text
+
+
+def test_static_warning_renders_in_style_section():
+    """Severity=warning items go to the style tips section, not blocking."""
+    card = build_feedback_card(
+        _task(),
+        ci_failures=[],
+        static_issues=[
+            StaticIssue(rule="style", message="新增 4 个 print()", severity="warning"),
+        ],
+        llm_issues=[],
+    )
+    text = "\n".join(e.get("content", "") for e in card["body"]["elements"])
+    assert "风格提示" in text
+    assert "print()" in text
+    # Should not appear in the blocking static section.
+    assert "静态规则违规" not in text

